app/assets/src/components/visualizations/heatmap/Heatmap.js

```diff
@@ -400,6 +439,165 @@ export default class Heatmap {
           this.options.spacing}
       )`
     );
+
+    // Set up space bar+click to pan behavior.
+    d3
+      .select("body")
+      .on("keydown", () => {
+        if (d3.event.code === "Space") {
+          this.spacePressed = true;
+        }
+      })
+      .on("keyup", () => {
+        if (d3.event.code === "Space") {
+          this.spacePressed = false;
+        }
+      });
+
+    this.svg
+      .on("mousedown", () => {
+        this.mouseDown = true;
+        if (this.spacePressed) {
+          this.mouseX = d3.event.clientX;
+          this.mouseY = d3.event.clientY;
+          d3.event.preventDefault();
+        }
+      })
+      .on("mousemove", this.drag.bind(this))
+      .on("mouseup", () => {
+        this.mouseDown = false;
+      });
+
+    // Set up scrolling behavior.
+    this.gCells.on("wheel.zoom", this.scroll.bind(this));
+  }
+
+  drag() {
+    if (this.mouseDown && this.spacePressed) {
+      this.pan(d3.event.clientX - this.mouseX, d3.event.clientY - this.mouseY);
+      this.mouseX = d3.event.clientX;
+      this.mouseY = d3.event.clientY;
+    }
+  }
+
+  scroll() {
+    this.pan(d3.event.wheelDeltaX, d3.event.wheelDeltaY);
+    d3.event.stopPropagation();
+  }
+
+  pan(deltaX, deltaY) {
+    // Define the scrolling boundaries for the svg.
+    // Upper limits are determined by the difference between the container and svg sizes,
+    // scaled by the zoom factor.
+    let containerWidth = this.container[0][0].offsetWidth,
+      containerHeight = this.container[0][0].offsetHeight;
+    let xScrollMax =
```

Comment:
Nit: I think this is actually the xScrollMin since it's usually a negative number.

Reply:
Also, I think this should take the `marginLeft` into account as well. Maybe it's `containerWidth - this.svg.atr("width") - this.options.marginLeft - this.options.marginRight`? Otherwise, as you zoom in more and more, I think there will be a bigger and bigger gap at the bottom and right when you try to scroll all the way.